### PR TITLE
[IOTDB-1391]  Add a new Aggregation function ext

### PR DESCRIPTION
**Pull request by lyssom**
## Description
Add a new aggregation function: ext

### Content3 ...

<hr>

This PR has:
- [ ] been self-reviewed.
    - [ ] concurrent read
    - [ ] concurrent write
    - [ ] concurrent read and write 
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. 
- [ ] added or updated version, __license__, or notice information
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious 
  for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold 
  for code coverage.
- [ ] added integration tests.
- [ ] been tested in a test IoTDB cluster.

<hr>

### CODE REVIEW

**Comment by LeiRui** [comment]
Could you give an explanation of what `ext` does?
I used to think that it functions as the max absolute function,
but the test example in the `AggregateResultTest.java` says that if we have two values `1` and `-2`, `ext` will return `-2`. So I'm confused.

File: server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+  public ExtremeAggrResult(TSDataType dataType) {
+    super(dataType, AggregationType.EXTREME);
+    reset();
+  }
+
+  public Object getAbsValue(Object v) {
+    double doubleValue;
+    float floatValue;
+    int intValue;
+    long longValue;
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java:45]: No need to define these

File: server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+    switch (resultDataType) {
+      case DOUBLE:
+        doubleValue = (Double) v;
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java:49]:
```suggestion
        double doubleValue = (Double) v;
```

File: server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+        return Math.abs(doubleValue);
+      case FLOAT:
+        floatValue = (Float) v;
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java:52]:
```suggestion
        float floatValue = (Float) v;
```

File: server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+    }
+  }
+
+  public Comparable<Object> compareAbsVal(Comparable<Object> minVal, Comparable<Object> maxVal) {
+    Comparable<Object> extVal;
+
+    Comparable<Object> absMaxVal = (Comparable<Object>) getAbsValue(maxVal);
+    Comparable<Object> absMinVal = (Comparable<Object>) getAbsValue(minVal);
+
+    if (absMaxVal.compareTo(absMinVal) >= 0) {
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java:71]: If the maxVal is 2, minVal is -2, here you choose the maxVal 2 as your defined extremed value which is inconsistent with the following.

File: server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+      if (extVal == null) {
+        extVal = currentValue;
+      } else {
+        Comparable<Object> absExtVal = (Comparable<Object>) getAbsValue(extVal);
+        if (absExtVal.compareTo(absCurrentValue) < 0) {
+          extVal = currentValue;
+        } else if (absExtVal.compareTo(absCurrentValue) == 0) {
+          if (extVal.compareTo(currentValue) > 0) {
+            extVal = currentValue;
+          }
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java:117]: Here, if we firstly meet the `2`, and you set `extVal` to `2` and then we meet `-2`, you will set `extVal` to `-2` which is inconsistent with the above.

File: server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+      if (currentValue != null) {
+        if (extVal == null) {
+          extVal = currentValue;
+        } else {
+          Comparable<Object> absExtVal = (Comparable<Object>) getAbsValue(extVal);
+          if (absExtVal.compareTo(absCurrentValue) < 0) {
+            extVal = currentValue;
+          } else if (absExtVal.compareTo(absCurrentValue) == 0
+              && extVal.compareTo(currentValue) > 0) {
+            extVal = currentValue;
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java:144]: same as above.

File: server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java:167]: same as above

File: server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+  @Override
+  public void merge(AggregateResult another) {
+    this.updateResult((Comparable<Object>) another.getResult());
+  }
+
+  @Override
+  protected void deserializeSpecificFields(ByteBuffer buffer) {}
+
+  @Override
+  protected void serializeSpecificFields(OutputStream outputStream) {}
```
Comment (JackieTien97) [inline server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java:189]: Try to override these two methods, you can refer to `LastValueAggrResult`

**Review by JackieTien97 — CHANGES_REQUESTED** [review]
Besides UT, you should add some ITs. Also, don't forget to add some docs in `ExtremeAggrResult` class to describe what does the extreme value mean, especially the corner case, such as if the max value and min value has the same abs value, which you should choose as the extreme value.

File: server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+  public Object getAbsValue(Object v) {
+    double doubleValue;
+    float floatValue;
+    int intValue;
+    long longValue;
+
+    switch (resultDataType) {
+      case DOUBLE:
+        doubleValue = (Double) v;
+        return Math.abs(doubleValue);
```
Comment (Alima777) [inline server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java:50]: There is no need to use global varaible.
```suggestion

    switch (resultDataType) {
      case DOUBLE:
        return Math.abs((Double) v);
```

File: server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+    Comparable<Object> absMaxVal = (Comparable<Object>) getAbsValue(maxVal);
+    Comparable<Object> absMinVal = (Comparable<Object>) getAbsValue(minVal);
+
+    if (absMaxVal.compareTo(absMinVal) >= 0) {
+      extVal = maxVal;
+    } else {
+      extVal = minVal;
+    }
+    return extVal;
+  }
```
Comment (Alima777) [inline server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java:77]:
```suggestion
  public Comparable<Object> compareAbsVal(Comparable<Object> minVal, Comparable<Object> maxVal) {
  Comparable<Object> absMaxVal = (Comparable<Object>) getAbsValue(maxVal);
    Comparable<Object> absMinVal = (Comparable<Object>) getAbsValue(minVal);
    return absMaxVal.compare(absMinVal) >= 0 ? maxVal : minval;
```

File: server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+          if (extVal.compareTo(currentValue) > 0) {
+            extVal = currentValue;
+          }
+        }
+      }
+
+      dataInThisPage.next();
+    }
+    updateResult(extVal);
+  }
```
Comment (Alima777) [inline server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java:124]: Please see `MaxValueAggrResult.java` to rewrite this method. Your code logic is confusing.

**Review by Alima777 — CHANGES_REQUESTED** [review]
Please see 'MaxValueAggrResult` to optimize your code logic. And Add integration test please.

File: server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java (hunk 4 above)
Comment (lyssom) [reply server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java]: sorry, I wrote wrong. It isfixed

File: server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java (hunk 3 above)
Comment (lyssom) [reply server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java]: ok，fixed

File: server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java (hunk 9 above)
Comment (lyssom) [reply server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java]: ok

File: server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java (hunk 4 above)
Comment (lyssom) [reply server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java]: I added IT.I will add docs later

File: server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+        extVal = currentValue;
+      } else {
+        Comparable<Object> absExtVal = (Comparable<Object>) getAbsValue(extVal);
+        if (absExtVal.compareTo(absCurrentValue) < 0) {
+          extVal = currentValue;
+        } else if (absExtVal.compareTo(absCurrentValue) == 0
+            && extVal.compareTo(currentValue) < 0) {
+          extVal = currentValue;
+        }
+      }
```
Comment (Alima777) [inline server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java:104]: Just use UpdateResult(extVal)

File: server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+          Comparable<Object> absExtVal = (Comparable<Object>) getAbsValue(extVal);
+          if (absExtVal.compareTo(absCurrentValue) < 0) {
+            extVal = currentValue;
+          } else if (absExtVal.compareTo(absCurrentValue) == 0
+              && extVal.compareTo(currentValue) < 0) {
+            extVal = currentValue;
+          }
+        }
+      }
+    }
```
Comment (Alima777) [inline server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java:134]: The same.

File: server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+    Comparable<Object> absCandidateResult = (Comparable<Object>) getAbsValue(getValue());
+
+    if (!hasCandidateResult()) {
+      setValue(extVal);
+    } else if (absExtVal.compareTo(absCandidateResult) > 0) {
+      setValue(extVal);
+    } else if (absExtVal.compareTo(absCandidateResult) == 0
+        && extVal.compareTo(candidateResult) > 0) {
+      setValue(extVal);
+    }
```
Comment (Alima777) [inline server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java:198]: Use `or` instead of else if

**Review by Alima777 — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/apache/iotdb/db/qp/constant/SQLConstant.java [hunk 15]
```diff
@@ -61,6 +61,8 @@ private SQLConstant() {
   public static final String MAX_VALUE = "max_value";
   public static final String MIN_VALUE = "min_value";
 
+  public static final String EXTREME = "ext";
```
Comment (Alima777) [inline server/src/main/java/org/apache/iotdb/db/qp/constant/SQLConstant.java:64]: Use `EXT`  please, `Extreme` = "ext" may cause confusion.

File: server/src/test/java/org/apache/iotdb/db/integration/aggregation/IoTDBAggregationIT.java [hunk 16]
```diff
@@ -694,6 +695,63 @@ public void maxminValueTest() {
...
+  @Test
+  public void extValueTest() {
+    String[] retArray = new String[] {"0,8499", "0,2499"};
+    try (Connection connection =
+            DriverManager.getConnection("jdbc:iotdb://127.0.0.1:6667/", "root", "root");
+        Statement statement = connection.createStatement()) {
+
+      boolean hasResultSet =
+          statement.execute(
+              "SELECT ext(s0) " + "FROM root.vehicle.d0 WHERE time >= 100 AND time < 9000");
```
Comment (Alima777) [inline server/src/test/java/org/apache/iotdb/db/integration/aggregation/IoTDBAggregationIT.java:707]: Why a "+"

**Review by Alima777 — CHANGES_REQUESTED** [review]

File: server/src/test/java/org/apache/iotdb/db/integration/aggregation/IoTDBAggregationIT.java (hunk 16 above)
Comment (lyssom) [reply server/src/test/java/org/apache/iotdb/db/integration/aggregation/IoTDBAggregationIT.java]: sorry, wrote wrong. fixed

File: server/src/main/java/org/apache/iotdb/db/qp/constant/SQLConstant.java (hunk 15 above)
Comment (lyssom) [reply server/src/main/java/org/apache/iotdb/db/qp/constant/SQLConstant.java]: ok

File: server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java (hunk 14 above)
Comment (lyssom) [reply server/src/main/java/org/apache/iotdb/db/query/aggregation/impl/ExtremeAggrResult.java]: ok

**Review by Alima777 — APPROVED** [review]
LGTM :D

**Pull request merged**